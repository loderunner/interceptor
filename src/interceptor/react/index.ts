export { Provider as InterceptProvider, useIntercepts } from './intercept'
