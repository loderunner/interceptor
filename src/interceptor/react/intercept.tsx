import * as React from 'react'
import { createContext, useContext, useState } from 'react'

import * as intercept from '../intercept'

interface InterceptContextType {
  intercepts: intercept.Intercept[]
  addIntercept: (pattern: string) => void
  removeIntercept: (inter: intercept.Intercept) => void
}

const InterceptContext = createContext<InterceptContextType>({
  intercepts: [],
  addIntercept: () => {},
  removeIntercept: () => {},
})

interface Props {
  children: React.ReactNode
}

export const Provider = ({ children }: Props) => {
  const [intercepts, setIntercepts] = useState([...intercept.intercepts])

  const addIntercept = (pattern: string) => {
    intercept.addIntercept({ pattern, enabled: true })
    setIntercepts([...intercept.intercepts])
  }
  const removeIntercept = (inter: intercept.Intercept) => {
    intercept.removeIntercept(inter)
    setIntercepts([...intercept.intercepts])
  }

  return (
    <InterceptContext.Provider
      value={{ intercepts, addIntercept, removeIntercept }}
    >
      {children}
    </InterceptContext.Provider>
  )
}

export const useIntercepts = () => useContext(InterceptContext)
