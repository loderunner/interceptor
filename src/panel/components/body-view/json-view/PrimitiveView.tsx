import * as React from 'react'
import { useCallback, useLayoutEffect, useMemo, useRef, useState } from 'react'

import ModalInput from '../../ModalInput'

import type { JSONPrimitive } from '../JSON'
import type { SyntheticEvent } from 'react'

interface Props {
  value?: JSONPrimitive
  onChange?: (value: JSONPrimitive) => void
  editingInitial?: boolean
}

const PrimitiveView = ({ value, onChange, editingInitial = false }: Props) => {
  const preRef = useRef<HTMLPreElement>(null)
  const [editing, setEditing] = useState(false)

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useLayoutEffect(() => setEditing(editingInitial), [])

  const onDoubleClick = useCallback((e: SyntheticEvent) => {
    if (!preRef.current) {
      return
    }
    e.stopPropagation()
    setEditing(true)
  }, [])

  const onChangeInput = useCallback(
    (value) => {
      if (onChange !== undefined) {
        onChange(JSON.parse(value))
      }
      setEditing(false)
    },
    [onChange]
  )

  const onCancelInput = useCallback(() => setEditing(false), [])

  const textColor = useMemo(() => {
    switch (typeof value) {
      case 'undefined':
        return ''
      case 'boolean':
        return 'text-green-500'
      case 'number':
        return 'text-rose-500'
      case 'string':
        return 'text-sky-400'
      case 'object': {
        if (value === null) {
          return 'text-purple-500'
        }
      }
      // fall through
      default:
        throw new Error('invalid JSON value')
    }
  }, [value])

  const valueString = useMemo(
    () => (value !== undefined ? JSON.stringify(value) : ''),
    [value]
  )

  return (
    <div className="flex-auto">
      <pre
        className={`whitespace-pre-wrap break-all ${textColor}`}
        ref={preRef}
        onDoubleClick={onDoubleClick}
      >
        {valueString}
      </pre>
      {editing && preRef.current ? (
        <ModalInput
          element={preRef.current}
          value={valueString}
          onChange={onChangeInput}
          onCancel={onCancelInput}
        />
      ) : null}
    </div>
  )
}

export default PrimitiveView
