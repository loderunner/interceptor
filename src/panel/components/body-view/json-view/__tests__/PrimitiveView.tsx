import { cleanup, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as React from 'react'
import '@testing-library/jest-dom'

import PrimitiveView from '../PrimitiveView'

describe('[JSON.PrimitiveView]', () => {
  afterEach(() => {
    cleanup()
  })

  it('should match null snapshot', () => {
    const { container } = render(<PrimitiveView value={null} />)
    expect(container).toMatchSnapshot()
  })

  it('should match boolean snapshot', () => {
    const { container } = render(<PrimitiveView value={true} />)
    expect(container).toMatchSnapshot()
  })

  it('should match number snapshot', () => {
    const { container } = render(<PrimitiveView value={11.38} />)
    expect(container).toMatchSnapshot()
  })

  it('should match string snapshot', () => {
    const { container } = render(<PrimitiveView value="Hello World" />)
    expect(container).toMatchSnapshot()
  })

  it('should open text field on double-click', async () => {
    const user = userEvent.setup()
    const value = 'Hello World!'
    const { getByText, queryByRole } = render(<PrimitiveView value={value} />)

    const pre = getByText(JSON.stringify(value))

    await user.dblClick(pre)

    const textbox = queryByRole('textbox')
    expect(textbox).not.toBeNull()
    expect(textbox).toHaveValue(JSON.stringify(value))
  })

  it('should call onChange on close', async () => {
    const user = userEvent.setup()
    const value = 'Hello World!'
    const onChange = jest.fn()
    const { getByText } = render(
      <PrimitiveView value={value} onChange={onChange} />
    )

    const pre = getByText(JSON.stringify(value))

    await user.dblClick(pre)
    await user.keyboard('true{Enter}')

    expect(onChange).toHaveBeenCalledWith(true)
  })
})
