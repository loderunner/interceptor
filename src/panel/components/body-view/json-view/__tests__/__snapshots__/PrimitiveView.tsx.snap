// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[JSON.PrimitiveView] should match boolean snapshot 1`] = `
<div>
  <div
    class="flex-auto"
  >
    <pre
      class="whitespace-pre-wrap break-all text-green-500"
    >
      true
    </pre>
  </div>
</div>
`;

exports[`[JSON.PrimitiveView] should match null snapshot 1`] = `
<div>
  <div
    class="flex-auto"
  >
    <pre
      class="whitespace-pre-wrap break-all text-purple-500"
    >
      null
    </pre>
  </div>
</div>
`;

exports[`[JSON.PrimitiveView] should match number snapshot 1`] = `
<div>
  <div
    class="flex-auto"
  >
    <pre
      class="whitespace-pre-wrap break-all text-rose-500"
    >
      11.38
    </pre>
  </div>
</div>
`;

exports[`[JSON.PrimitiveView] should match string snapshot 1`] = `
<div>
  <div
    class="flex-auto"
  >
    <pre
      class="whitespace-pre-wrap break-all text-sky-400"
    >
      "Hello World"
    </pre>
  </div>
</div>
`;
